objects/src/accounts/account_id.rs

```diff
@@ -0,0 +1,200 @@
+use super::{AccountError, Felt, Hasher, StarkField, Word, ZERO};
+use core::fmt;
+
+// ACCOUNT ID
+// ================================================================================================
+
+/// Unique identifier of an account.
+///
+/// Account ID consists of 3 field elements (24 bytes). These field elements uniquely identify a
+/// single account and also specify the type of the underlying account. Specifically:
+/// - If the least significant 32 bits of the 3rd element are all ZEROs, the account is a
+///   fungible asset faucet (i.e., it can issue fungible assets).
+/// - If the least significant 32 bits of the 3rd element are set to 2^31 (i.e ONE followed by 31
+///   ZEROs), the account is a non-fungible asset faucet (i.e., it can issue non-fungible assets).
+/// - Otherwise, the account is a regular account.
+///
+/// Additionally, account IDs have the following properties
+/// - For fungible asset faucets account IDs are guaranteed to start with ONE.
+/// - For regular accounts, the last 3 bytes of the ID are guaranteed to be all ZEROs.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct AccountId(Felt, Felt, Felt);
+
+impl AccountId {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+    pub const FUNGIBLE_FAUCET_TAG: u32 = 0;
+    pub const NON_FUNGIBLE_FAUCET_TAG: u32 = 1 << 31;
+
+    pub const MIN_TRAILING_ZEROS: u32 = 24;
+
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns a new account ID derived from the specified seed.
+    ///
+    /// The account ID is computed by hashing the seed and using 3 elements of the result to form
+    /// the ID. Specifically we are take elements 0, 1, and 3, omitting element 2. We omit element
+    /// 2 because unlike elements 0 and 3, it has no special structure which we need to carry over
+    /// into the derived account ID. Element 1 could have been omitted just as well.
+    ///
+    /// # Errors
+    /// Returns an error if the resulting account ID does not comply with account ID rules.
+    pub fn new(seed: Word) -> Result<Self, AccountError> {
+        // hash the seed and build the ID from the 1st, 2nd, and 4th elements of the result
+        let hash = Hasher::hash_elements(&seed);
```

Comment:
I'd rather not expose keep the hash function as implementation detail so that the user doesn't need to know which hash function to use for instantiating IDs.

Reply:
That makes sense - also if this is defined protocol level